v2transport/transport.go

```diff
@@ -0,0 +1,625 @@
+	// the BitcoinNet's "magic" bytes.
+	salt := []byte("bitcoin_v2_shared_secret")
+
+	var magic [4]byte
+	binary.LittleEndian.PutUint32(magic[:], uint32(net))
+	salt = append(salt, magic[:]...)
+
+	// Use the hkdf Extract function to generate a pseudo-random key.
+	prk := hkdf.Extract(sha256.New, ecdh_secret, salt)
+
+	// Use the hkdf Expand function with info set to "session_id" to generate a
+	// unique sessionID.
+	sessionInfo := []byte("session_id")
+	sessionReader := hkdf.Expand(sha256.New, prk, sessionInfo)
+	_, err := sessionReader.Read(p.sessionID)
+	if err != nil {
+		return err
+	}
+
+	// Use the Expand operation to generate packet and packet-length encryption
+	// ciphers.
+	initiatorLInfo := []byte("initiator_L")
+	initiatorLReader := hkdf.Expand(sha256.New, prk, initiatorLInfo)
+	_, err = initiatorLReader.Read(p.initiatorL)
+	if err != nil {
+		return err
+	}
+
+	initiatorPInfo := []byte("initiator_P")
+	initiatorPReader := hkdf.Expand(sha256.New, prk, initiatorPInfo)
+	_, err = initiatorPReader.Read(p.initiatorP)
+	if err != nil {
+		return err
+	}
+
+	responderLInfo := []byte("responder_L")
+	responderLReader := hkdf.Expand(sha256.New, prk, responderLInfo)
+	_, err = responderLReader.Read(p.responderL)
+	if err != nil {
+		return err
+	}
+
+	responderPInfo := []byte("responder_P")
+	responderPReader := hkdf.Expand(sha256.New, prk, responderPInfo)
+	_, err = responderPReader.Read(p.responderP)
+	if err != nil {
+		return err
+	}
+
+	// Create the garbage terminators that each side will use.
+	garbageInfo := []byte("garbage_terminators")
+	garbageReader := hkdf.Expand(sha256.New, prk, garbageInfo)
+	garbageTerminators := make([]byte, 32)
+	_, err = garbageReader.Read(garbageTerminators)
+	if err != nil {
+		return err
+	}
+
+	initiatorGarbageTerminator := garbageTerminators[:garbageSize]
+	responderGarbageTerminator := garbageTerminators[garbageSize:]
+
+	if initiating {
+		p.sendL, err = NewFSChaCha20(p.initiatorL)
+		if err != nil {
+			return err
+		}
+
+		p.sendP, err = NewFSChaCha20Poly1305(p.initiatorP)
+		if err != nil {
+			return err
+		}
+
+		p.sendGarbageTerm = initiatorGarbageTerminator
+
+		p.recvL, err = NewFSChaCha20(p.responderL)
+		if err != nil {
+			return err
+		}
+
+		p.recvP, err = NewFSChaCha20Poly1305(p.responderP)
+		if err != nil {
+			return err
+		}
+
+		p.recvGarbageTerm = responderGarbageTerminator
+	} else {
+		p.sendL, err = NewFSChaCha20(p.responderL)
+		if err != nil {
+			return err
+		}
+
+		p.sendP, err = NewFSChaCha20Poly1305(p.responderP)
+		if err != nil {
+			return err
+		}
+
+		p.sendGarbageTerm = responderGarbageTerminator
+
+		p.recvL, err = NewFSChaCha20(p.initiatorL)
+		if err != nil {
+			return err
+		}
+
+		p.recvP, err = NewFSChaCha20Poly1305(p.initiatorP)
+		if err != nil {
+			return err
+		}
+
+		p.recvGarbageTerm = initiatorGarbageTerminator
+	}
+
+	// TODO:
+	//     To achieve forward secrecy we must wipe the key material used to initialize the ciphers:
```

Comment:
I think in go we can mainly w/ best effort basis:
  * use something like `mlock` to prevent the secret data from swapping to disk 
  * make functions to zero out bytes either by doing each byte, or setting the pointer to an empty value
      * optionally force a GC at this point 

Can be wroth checking if the crypto stuff in the stdlib does anything fancy here.

Reply:
Looks like there's this issue: https://github.com/golang/go/issues/21865